README
Term::RouterCLI is Copyright (C) 2011, Bret Jordan
==================================================
HOSTED AT
    http://ciscocli.sourceforge.net


MAJOR FEATURES
    The major features of Term::RouterCLI are as follows:
    Extensible tree based parsing engine for a command line interface
    Modular design
    TAB and ? mark completion
    Recursive completion
    Recursive command interpolation
    Hidden commands and command trees
    Authenticated commands and command trees
    Multi-user authentication
    Per user command histories
    Multiple authentication servers
    Audit log support
    International support for all command descriptions, help documentation, and argument help
    Hardware interaction framework as an example 
    Fully working cli.pl example
    Apache style configuration file
    Separate vendor and user parts to the configuration file
    Uses log4perl for debugging


INSTALLATION
    To install this module type the following:
    perl Makefile.PL
    make
    make test
    make install


DEPENDENCIES
    This module requires these other modules and libraries:
    Config::General (2.50)          [License = Perl]
    Digest::SHA (5.62)              [License = Perl]
    Env (1.00)                      [License = Perl]
    FileHandle (2.02)               [License = Perl]
    Log:Log4perl (1.33)             [License = Perl]
    parent (0.221)                  [License = Perl]
    POSIX (1.17)                    [License = Perl]
    Sys::Syslog (0.29)              [License = Perl]
    Term::ReadKey (2.30)            [License = Opensource]
    Term::ReadLine::Gnu (1.20)      [License = Perl]
    Test::More (0.98)               [License = Perl]
    Test::Output (1.01)             [License = Perl]
    Text::Shellwords::Cursor (0.81) [License = MIT]
    Ethtool (http://www.kernel.org/pub/software/network/ethtool/)


NOTES
    The tarball has a fully working example 
    
    I use a modified version of Ethtool so as to get rid of some STDERR messages.
    
    To find out what version of a module you have installed run something like the following
        perl -MDigest::SHA -e 'print "$Digest::SHA::VERSION\n"'


LICENCE INFORMATION
    See the LICENSE file included with this package for license details. 


AUTHOR
    Bret Jordan, jordan at open1x littledot org, jordan2175 at gmail littledot com


COPYRIGHT
    Copyright (C) 2011 by Bret Jordan all rights reserved



Changes
Release Notes:
v0.99_16 - Beta 6 Release ()
    Changes
        Setup an actual configuration object for Running and Startup Configs
        Removed the need to pass the config object in to other classes
        Cleaned up some of the need to pass the parent object around
        Fixed some of the debug statements
    
v0.99_15 - Beta 5 Release (2011-09-14)
    Changes
        Implemented Log::Log4perl
        Various bug fixes

v0.99_14 - Beta 4 Release (Non released version)
    Changes
        Minor fixes and tweaks

v0.99_13 - Beta 3 Release (2011-07-26)
    Changes
        Renamed project from Term::CiscoCLI to Term::RouterCLI

v0.99_12 - Beta 2 Release (2011-07-25)
    Changes
        Fixed issues when prompting for things like a password or username
        Added support for multi-user authentication 
        Cleaned up support for authentication of commands or command trees in general
        Added support to the audit log to track users and TTYs
        Change requirement of perl 5.10.1 to 5.8.8 and documented what is needed for 5.8.8 to work
    Known Issues
        Hidden command settings is not inherited
        There is no built in methods for changing passwords or creating users, not yet
        You can only use one user database right now
        RADIUS support is not yet working
        
v0.99_11 - Beta 1 Release (2011-07-12)
    Changes
        Lots and lots of fixes and enhancements
        Added a hardware class layer
        Added support for authenticating commands
        Got "config t" mode working in the examples
        Added support for showing interface stats
        Built config support for MOTDs
        Added vendor specific parts to the Config file
        Modified the interaction with the language object
        Added encrypted password support
        
v0.99_04 - Alpha 4 Release (2011-04-27)
    Changes
        Created a new Log class
        Rewrote the history and auditlog classes
        Fixed the default command
        Added more public convenience functions to improve generalization
        Rewrote the language classes

v0.99_03 - Alpha 3 Release (2011-04-23)
    Changes
        Enhanced unit tests
        Reduced "args" directive down to just code execution 
        Removed unused methods
        Added show subcommands to help
        More bug fixes
    Known Issues
        Max and Min args still do not work right during completion
        The default command is still untested
        Hidden command settings is not inherited
        The code needs to be generalized so that you can set history, audit log files, and other options

v0.99_02 - Alpha 2 Release (2011-04-20)
    Changes
        Added module requirements to the Makefile.PL
        Lots of cleanup and bug fixes

v0.99_01 - Initial Alpha Release (2011-04-18)
    Differences from v0.98 of Term::ShellUI
        There is currently no file or directory completion
        There is currently no code based prompts
        There is currently no array based "args" directive option
        The "proc" directive has been deprecated
        The "method" directive has been renamed to "code"
    New Features
        Abbreviated commands work such as "sh int eth" for "show interface eth0"
        Configuration file support for storing changes between sessions
        Basic internationalization support for help menu and command summaries
        Hidden command trees
        Authenticated command trees
        Audit log support
        Support for Syslog logging 
        Objectized large parts of the code
    Known Issues
        Far to many to list


Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME            => 'Term::RouterCLI',
    VERSION_FROM    => 'lib/Term/RouterCLI.pm', # finds $VERSION
    ABSTRACT        => 'Classes for building a Router (Standford) style CLI',
    PREREQ_PM       => {
            'Config::General'           => 2.50,
            'Digest::SHA'               => 5.62,
            'Env'                       => 1.00,
            'FileHandle'                => 2.02,
            'parent'                    => 0.221,
            'POSIX'                     => 1.17,
            'Sys::Syslog'               => 0.29,
            'Term::ReadKey'             => 2.30,
            'Term::ReadLine::Gnu'       => 1.20,
            'Test::More'                => 0.98,
            'Test::Output'              => 1.01,
            'Text::Shellwords::Cursor'  => 0.81,
    },
    AUTHOR          => 'Bret Jordan',
);
